### PR TITLE
Update project to use loopback-3.x

### PR DESCRIPTION
**Pull request by gunjpan**
Update to use `loopback-3.x` and `strong-remoting 3.x`

### CODE REVIEW

**Comment by gunjpan** [comment]
**Update 29/11**
- Blocked until new version of `strong-remoting` is released to npmjs.org

File: LoopBackTests/server/index.js
Comment (bajtos) [inline LoopBackTests/server/index.js]: nitpick (feel free to ignore): we usually use `db` as the name of the main datasource.

File: LoopBackTests/server/index.js
Comment (bajtos) [inline LoopBackTests/server/index.js]: AFAIK, the `GET` method explicitly forbids non-empty request bodies, therefore there will be no `form` to get the value from.

I assume the tests are passing for you, which may mean that `source: 'form'` is ignored, or that the test is sending a `GET` request with a non-empty body. Could you please investigate?

File: LoopBackTests/server/index.js
Comment (bajtos) [inline LoopBackTests/server/index.js]: This should be automatically handled by `app.enableAuth({ dataSource: 'Memory' });`.

File: LoopBackTests/server/index.js
Comment (bajtos) [inline LoopBackTests/server/index.js]: Since we are not calling `autoAttach`, the `mail` datasource should not be needed at all now.

File: SLRemoting/SLObject.m
Comment (bajtos) [inline SLRemoting/SLObject.m]: Is there a way how to support both LB2.x and LB3.x servers with a single SDK client? I am concerned that with this change in place, users won't be able to use the latest SDK version with their LB2.x apps.

Also I don't recall changing the format of date arguments in strong-remoting, why is this fix needed in the first place? Is this perhaps a side-effect of my coercion refactor that I did not discover?

**Review by bajtos — CHANGES_REQUESTED** [review]

File: LoopBackTests/server/index.js
Comment (gunjpan) [inline LoopBackTests/server/index.js]: Yes, it occurred to me but kept as is to be in sync with existing code

File: SLRemoting/SLObject.m
Comment (gunjpan) [inline SLRemoting/SLObject.m]: >Is this perhaps a side-effect of my coercion refactor that I did not discover?

There was some coercion issue, don't remember ATM. Will update you with the findings.

Comment (gunjpan) [inline SLRemoting/SLObject.m]: @bajtos :
Update: 
` value = @{ @"$type": @"date", @"$data": jsonString };`

gets converted to following URL string, which is not coerced correctly 
`http://localhost:3000/widgets/testDate?date[%24data]=1970-01-01T00%3A02%3A03.000Z&date[%24type]=date`

File: LoopBackTests/server/index.js
Comment (bajtos) [inline LoopBackTests/server/index.js]: Fair enough.

File: SLRemoting/SLObject.m
Comment (bajtos) [inline SLRemoting/SLObject.m]: I see. Could you please check what happens when you use the new simplified encoding `testDate?date=1970-01-01T00%3A02%3A03.000Z` when calling a LoopBack 2.x server? Will the server correctly convert the value to a Date instance?

I think this is a subtle but also an important breaking change, could you please add a note to both Release Notes and Migration Guide for 3.0?

**Comment by rmg** [comment]
@slnode test please

**Comment by gunjpan** [comment]
@slnode test please

File: SLRemoting/SLObject.m
Comment (bajtos) [inline SLRemoting/SLObject.m]: > I think this is a subtle but also an important breaking change, could you please add a note to both Release Notes and Migration Guide for 3.0?

I mean that Release Notes should mention that LoopBack 3.x not longer recognizes nested query arguments encoded using `{$type,$data}` schema and show what's the new format. The Migration Guide should have a short section telling users to upgrade their client apps.

**Comment by bajtos** [comment]
> Could you please check what happens when you use the new simplified encoding `testDate?date=1970-01-01T00%3A02%3A03.000Z` when calling a LoopBack 2.x server? Will the server correctly convert the value to a Date instance?

@gunjpan have you managed to trie the above? What was the result?

**Comment by gunjpan** [comment]
@bajtos : 
>I see. Could you please check what happens when you use the new simplified encoding testDate?date=1970-01-01T00%3A02%3A03.000Z when calling a LoopBack 2.x server? Will the server correctly convert the value to a Date instance?

It worked fine for LB 2.x server, IIRC, it converts the value to Date instance, need to verify one more time.

**Comment by igordeoliveirasa** [comment]
any news? i'd like to use your version @gunjpan

**Comment by bajtos** [comment]
@igordeoliveirasa I am afraid we are no longer actively maintaining the project. Would you perhaps like to step up to become a maintainer yourself? If yes, then please leave a comment in https://github.com/strongloop-community/loopback-sdk-ios/issues/136

**Pull request closed without merging**
